presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp

```diff
@@ -117,68 +201,132 @@ RowVectorPtr BroadcastFileWriter::fileStats() {
            std::move(numRowsVector)}));
 }
 
-void BroadcastFileWriter::initializeWriteFile() {
-  if (!writeFile_) {
-    LOG(INFO) << "Opening broadcast file for write: " << filename_;
-    writeFile_ = fileSystem_->openFileForWrite(filename_);
-  }
-}
-
-void BroadcastFileWriter::write(const RowVectorPtr& rowVector) {
-  auto numRows = rowVector->size();
-  if (numRows == 0) {
-    return;
-  }
-
-  initializeWriteFile();
-  numRows_ += numRows;
-  const IndexRange allRows{0, numRows};
-
-  auto arena = std::make_unique<StreamArena>(pool_);
-  auto serializer =
-      serde_->createIterativeSerializer(inputType_, numRows, arena.get());
-
-  serializer->append(rowVector, folly::Range(&allRows, 1));
-  maxSerializedSize_ += serializer->maxSerializedSize();
-  IOBufOutputStream out(*pool_);
-  serializer->flush(&out);
-  auto iobuf = out.getIOBuf();
-  for (auto& range : *iobuf) {
-    writeFile_->append(std::string_view(
-        reinterpret_cast<const char*>(range.data()), range.size()));
-  }
-  writeFile_->flush();
-}
-
 BroadcastFileReader::BroadcastFileReader(
     std::unique_ptr<BroadcastFileInfo>& broadcastFileInfo,
     std::shared_ptr<velox::filesystems::FileSystem> fileSystem,
     velox::memory::MemoryPool* pool)
-    : broadcastFileInfo_(std::move(broadcastFileInfo)),
-      fileSystem_(fileSystem),
-      hasData_(true),
-      numBytes_(0),
-      pool_(pool) {}
+    : pool_(pool), broadcastFileInfo_(std::move(broadcastFileInfo)) {
+  readFile_ = fileSystem->openFileForRead(broadcastFileInfo_->filePath_);
+  readFooter();
+}
 
 bool BroadcastFileReader::hasNext() {
-  return hasData_;
+  return numPagesRead_ < pageSizes_.size();
 }
 
 velox::BufferPtr BroadcastFileReader::next() {
   if (!hasNext()) {
     return nullptr;
   }
 
-  auto readFile = fileSystem_->openFileForRead(broadcastFileInfo_->filePath_);
-  auto buffer = AlignedBuffer::allocate<char>(readFile->size(), pool_, 0);
-  readFile->pread(0, readFile->size(), buffer->asMutable<char>());
-  numBytes_ += readFile->size();
-  hasData_ = false;
-  return buffer;
+  int64_t pageSize = pageSizes_[numPagesRead_];
```

Comment:
Can we use FileInputStream or reuse class SpillReadFile with necessary refactor to make it generic?